.github/workflows/pytest.yml

```diff
@@ -0,0 +1,192 @@
+          # Needed because release/non-release downloads are different structure
+          cd hse && if [ ! -f *.whl ]; then mv */*.whl ./; fi
+
+      - name: Download amalgam-lang-py
+        if: contains(inputs.upstream-details, 'amalgam-lang-py')
+        env:
+          GH_TOKEN: ${{ secrets.HOWSOAI_WORKFLOW_AUTOMATION_CLASSIC_TOKEN }}
+        run: |
+          run_type=$(echo "${{ inputs.upstream-details }}" | jq '".amalgam-lang-py"."run-type"')
+          run_id=$(echo "${{ inputs.upstream-details }}" | jq '".amalgam-lang-py"."run-id"')
+          gh $run_type download -D amlg -R "howsoai/amalgam-lang-py" -p "amalgam_lang-*-py3-none-any*" "$run_id"
+          # Needed because release/non-release downloads are different structure
+          cd amlg && if [ ! -f *.whl ]; then mv */*.whl ./; fi
+
+      - name: Download howso-validator-py
+        if: contains(inputs.upstream-details, 'howso-validator-py')
+        env:
+          GH_TOKEN: ${{ secrets.HOWSOAI_WORKFLOW_AUTOMATION_CLASSIC_TOKEN }}
+        run: |
+          run_type=$(echo "${{ inputs.upstream-details }}" | jq '".howso-validator-py"."run-type"')
+          run_id=$(echo "${{ inputs.upstream-details }}" | jq '".howso-validator-py"."run-id"')
+          gh $run_type download -D hsv -R "howsoai/howso-validator-py" -p "howso_validator-*-py3-none-any*" "$run_id"
+          # Needed because release/non-release downloads are different structure
+          cd hsv && if [ ! -f *.whl ]; then mv */*.whl ./; fi
+
+      - name: Download howso-validator-enterprise-py
+        if: contains(inputs.upstream-details, 'howso-validator-enterprise-py')
+        env:
+          GH_TOKEN: ${{ secrets.HOWSOAI_WORKFLOW_AUTOMATION_CLASSIC_TOKEN }}
+        run: |
+          run_type=$(echo "${{ inputs.upstream-details }}" | jq '".howso-validator-enterprise-py"."run-type"')
+          run_id=$(echo "${{ inputs.upstream-details }}" | jq '".howso-validator-enterprise-py"."run-id"')
+          gh $run_type download -D hsve -R "howsoai/howso-validator-enterprise-py" -p "howso_validator_enterprise-*-py3-none-any*" "$run_id"
+          # Needed because release/non-release downloads are different structure
+          cd hsve && if [ ! -f *.whl ]; then mv */*.whl ./; fi
+
+      - name: Download howso-synthesizer-py
+        if: contains(inputs.upstream-details, 'howso-synthesizer-py')
+        env:
+          GH_TOKEN: ${{ secrets.HOWSOAI_WORKFLOW_AUTOMATION_CLASSIC_TOKEN }}
+        run: |
+          run_type=$(echo "${{ inputs.upstream-details }}" | jq '".howso-synthesizer-py"."run-type"')
+          run_id=$(echo "${{ inputs.upstream-details }}" | jq '".howso-synthesizer-py"."run-id"')
+          gh $run_type download -D synth -R "howsoai/howso-synthesizer-py" -p "howso_synthesizer-*-py3-none-any*" "$run_id"
+          # Needed because release/non-release downloads are different structure
+          cd synth && if [ ! -f *.whl ]; then mv */*.whl ./; fi
```

Comment:
My gut feeling is all of these can be collapsed into a loop with a carefully created data dict up front containing changing info.